File: bs/binarySearch.go
package main

import "fmt"

func search(nums []int, target int) int {

	n := len(nums)
	fmt.Println(n)
	l := 0
	r := n - 1
	for l < r {
		mid := l + (r-l)/2
		if target < nums[mid] {
			r = mid - 1
		} else if target > nums[mid] {
			l = mid + 1
		} else {
			return mid
		}

	}

	return -1

}

// Input: nums = [-1,0,3,5,9,12], target = 9
// Output: 4
func main() {
	nums := []int{-1, 0, 3, 5, 9, 12}
	target := 9
	fmt.Println(search(nums, target))
}

File: bs/binarySearch.py
class Solution(object):
    def search(self, nums, target):
        n = len(nums)

        l = 0
        r = n-1
        l, r = 0, len(nums) - 1
        while l <= r:
            m = l + ((r - l) // 2)  # (l + r) // 2 can lead to overflow
            if nums[m] > target:
                r = m - 1
            elif nums[m] < target:
                l = m + 1
            else:
                print(m)
                # return m
        return -1
    
# Input: nums = [-1,0,3,5,9,12], target = 9
        
nums = [-1,0,3,5,9,12]
target = 9
print(Solution().search(nums,target ))